File: crates/suspect-codegen/src/kotlin_sdk/Json.kt
package __PACKAGE__

import java.math.BigInteger
import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction

/** Exact JSON values. Collections may be caller-owned; codecs snapshot and revalidate them. */
public sealed interface JsonValue

/** JSON null. Nullable native model fields instead use Kotlin `null`. */
public data object JsonNull : JsonValue

/** A decoded Unicode string; JSON I/O rejects unpaired surrogates.
 * @property value Decoded string, without Unicode normalization.
 */
public data class JsonString(/** Decoded Unicode string. */ public val value: String) : JsonValue

/** A JSON boolean, distinct from every number.
 * @property value Boolean value.
 */
public data class JsonBoolean(/** Boolean value. */ public val value: Boolean) : JsonValue

/** A JSON array, retaining every element.
 * @property values Elements in wire order. Do not mutate during a codec call.
 */
public data class JsonArray(/** Elements in wire order. */ public val values: List<JsonValue>) : JsonValue

/** A JSON object, retaining every decoded key.
 * @property values Members. Do not mutate during a codec call.
 */
public data class JsonObject(/** Exact decoded member names and values. */ public val values: Map<String, JsonValue>) : JsonValue

/** An exact JSON number with a symbolic, unbounded decimal exponent.
 *
 * [parse] preserves the original token. Equality, hashing, ordering and integer
 * conversion are mathematical: `1`, `1.0` and `1e0` compare equal. No binary
 * floating-point or BigDecimal scale conversion is performed.
 * @property token Original validated ASCII numeric token.
 */
public class JsonNumber private constructor(/** Original validated ASCII numeric token. */ public val token: String) : JsonValue, Comparable<JsonNumber> {
    internal val exact: ExactDecimal by lazy(LazyThreadSafetyMode.PUBLICATION) { ExactDecimal.parse(token) }

    /** Original spelling, suitable for exact JSON emission. */
    override fun toString(): String = token
    /** Exact mathematical equality, independent of spelling. */
    override fun equals(other: Any?): Boolean = other is JsonNumber && exact == other.exact
    /** Mathematical hash, including one hash for every spelling of zero. */
    override fun hashCode(): Int = exact.hashCode()
    /** Exact order without exponent-sized allocation. */
    override fun compareTo(other: JsonNumber): Int = exact.compareTo(other.exact)
    /** Whether this value has no fractional component. */
    public fun isInteger(): Boolean = exact.integral

    /** Convert a mathematical integer, expanding at most 4096 decimal digits.
     * @throws ArithmeticException If this number has a fractional component.
     * @throws JsonException If exponent expansion would exceed the finite limit.
     */
    public fun toBigIntegerExact(): BigInteger {
        val number = exact
        if (!number.integral) throw ArithmeticException("JSON number is not an integer")
        if (number.sign == 0) return BigInteger.ZERO
        if (number.exponent + BigInteger.valueOf(number.digits.length.toLong()) > BigInteger.valueOf(4096)) {
            throw JsonException("integer expansion exceeds 4096 digits", 0, JsonErrorKind.RESOURCE_LIMIT)
        }
        val magnitude = BigInteger(number.digits) * BigInteger.TEN.pow(number.exponent.toInt())
        return if (number.sign < 0) magnitude.negate() else magnitude
    }

    /** Convert exactly to Long, rejecting fractions, overflow and huge expansion. */
    public fun toLongExact(): Long = toBigIntegerExact().longValueExact()

    /** Exact-number factories. */
    public companion object {
        private val syntax = Regex("-?(?:0|[1-9][0-9]*)(?:\\.[0-9]+)?(?:[eE][+-]?[0-9]+)?")

        /** Parse a strict JSON number of at most 4096 ASCII bytes. */
        public fun parse(token: String): JsonNumber {
            if (token.length > 4096) throw JsonException("number exceeds 4096 bytes", 0, JsonErrorKind.RESOURCE_LIMIT)
            if (!syntax.matches(token)) throw JsonException("invalid JSON number", 0)
            return JsonNumber(token)
        }

        /** Represent a Kotlin integer exactly. */
        public fun of(value: Long): JsonNumber = JsonNumber(value.toString())
        /** Represent a BigInteger within the numeric-token limit exactly. */
        public fun of(value: BigInteger): JsonNumber = parse(value.toString())
    }
}

/** Stable JSON failure categories. */
public enum class JsonErrorKind {
    /** Malformed JSON syntax, Unicode or UTF-8. */ SYNTAX,
    /** A finite byte, number, depth, value or work limit was exhausted. */ RESOURCE_LIMIT,
}

/** A bounded JSON error. Messages never include input tokens or object keys.
 * @property offset UTF-16 position in text; zero for byte/size errors.
 * @property kind Syntax versus incomplete resource-limited work.
 */
public class JsonException(message: String, /** UTF-16 text offset, or zero for byte errors. */ public val offset: Int, /** Syntax versus resource exhaustion. */ public val kind: JsonErrorKind = JsonErrorKind.SYNTAX) :
    IllegalArgumentException("$message at offset $offset")

/** Finite JSON policy; callers may lower each ceiling, including to zero.
 * @property maxBytes Maximum input/output UTF-8 bytes, at most 4 MiB.
 * @property maxDepth Maximum nesting depth, at most 128.
 * @property maxValues Maximum visited values, at most 100000.
 * @property maxNumberBytes Maximum bytes in one numeric token, at most 4096.
 * @property maxWorkBytes Maximum scanned/copied text units, at most 16 MiB.
 */
public data class JsonLimits(
    /** Input/output UTF-8 byte ceiling. */
    public val maxBytes: Int = 4 * 1024 * 1024,
    /** Maximum nesting depth. */
    public val maxDepth: Int = 128,
    /** Maximum visited values. */
    public val maxValues: Int = 100_000,
    /** Maximum bytes in one exact numeric token. */
    public val maxNumberBytes: Int = 4096,
    /** Maximum scanned/copied text units. */
    public val maxWorkBytes: Int = 16 * 1024 * 1024,
) {
    init {
        require(maxBytes in 0..4194304 && maxDepth in 0..128 && maxValues in 0..100000
            && maxNumberBytes in 0..4096 && maxWorkBytes in 0..16777216) { "JSON limits must be within the finite profile" }
    }
}

internal class JsonBudget(
    val limits: JsonLimits,
    val checkpoint: () -> Unit,
    private var remaining: Int = limits.maxValues,
    private var work: Int = limits.maxWorkBytes,
) {
    fun value(depth: Int) {
        checkpoint()
        if (depth >= limits.maxDepth || remaining-- <= 0) resource("JSON nesting/value limit exceeded")
    }
    fun spend(amount: Int) {
        if (amount > work) resource("JSON work limit exceeded")
        work -= amount
        checkpoint()
    }
    fun resource(message: String): Nothing = throw JsonException(message, 0, JsonErrorKind.RESOURCE_LIMIT)
}

/** Strict, finite exact JSON I/O. Duplicate keys and trailing data are rejected. */
public object Json {
    internal const val MAX_BYTES: Int = 4 * 1024 * 1024
    internal const val MAX_PROGRAM_BYTES: Int = 16 * 1024 * 1024

    // Generated metadata has its own finite load budget, separate from API payloads.
    internal fun parseProgram(bytes: ByteArray): JsonValue {
        if (bytes.size > MAX_PROGRAM_BYTES) throw JsonException("program byte limit exceeded", 0, JsonErrorKind.RESOURCE_LIMIT)
        return Parser(decodeUtf8(bytes), JsonBudget(JsonLimits(), {}, 1_000_000, 4 * MAX_PROGRAM_BYTES)).document()
    }

    /** Parse a document, retaining exact numbers and distinct Unicode keys. */
    public fun parse(text: String, limits: JsonLimits = JsonLimits()): JsonValue = parseChecked(text, limits) {}
    /** Decode UTF-8 strictly, then parse a document. */
    public fun parse(bytes: ByteArray, limits: JsonLimits = JsonLimits()): JsonValue = parseChecked(bytes, limits) {}
    /** Encode deterministically, checking every reachable value and output byte. */
    public fun stringify(value: JsonValue, limits: JsonLimits = JsonLimits()): String = stringifyChecked(value, limits) {}

    internal fun parseChecked(text: String, limits: JsonLimits, checkpoint: () -> Unit): JsonValue {
        utf8Size(text, limits.maxBytes, checkpoint)
        return Parser(text, JsonBudget(limits, checkpoint)).document()
    }
    internal fun parseChecked(bytes: ByteArray, limits: JsonLimits, checkpoint: () -> Unit): JsonValue {
        checkpoint()
        if (bytes.size > limits.maxBytes) throw JsonException("JSON byte limit exceeded", 0, JsonErrorKind.RESOURCE_LIMIT)
        val text = decodeUtf8(bytes)
        checkpoint()
        return Parser(text, JsonBudget(limits, checkpoint)).document()
    }

    private fun decodeUtf8(bytes: ByteArray): String = try {
        Charsets.UTF_8.newDecoder().onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT).decode(ByteBuffer.wrap(bytes)).toString()
    } catch (_: java.nio.charset.CharacterCodingException) { throw JsonException("invalid UTF-8", 0) }
    internal fun stringifyChecked(value: JsonValue, limits: JsonLimits, checkpoint: () -> Unit): String {
        val writer = Writer(JsonBudget(limits, checkpoint))
        writer.value(value, 0)
        return writer.out.toString()
    }

    private class Parser(private val text: String, private val budget: JsonBudget) {
        private var at = 0
        private fun fail(message: String): Nothing = throw JsonException(message, at)
        private fun space() {
            val start = at
            while (at < text.length && text[at] in " \t\r\n") {
                at++
                if ((at - start) % 1024 == 0) budget.checkpoint()
            }
            budget.spend(at - start)
        }
        fun document(): JsonValue {
            val result = value(0)
            space()
            if (at != text.length) fail("trailing JSON data")
            return result
        }
        private fun value(depth: Int): JsonValue {
            budget.value(depth)
            space()
            if (at == text.length) fail("unexpected end of JSON")
            return when (text[at]) {
                'n' -> { literal("null"); JsonNull }
                't' -> { literal("true"); JsonBoolean(true) }
                'f' -> { literal("false"); JsonBoolean(false) }
                '"' -> JsonString(string())
                '[' -> {
                    at++; space()
                    val values = mutableListOf<JsonValue>()
                    if (!take(']')) {
                        do { values.add(value(depth + 1)); space() } while (take(','))
                        expect(']')
                    }
                    JsonArray(values.toList())
                }
                '{' -> {
                    at++; space()
                    val values = linkedMapOf<String, JsonValue>()
                    if (!take('}')) {
                        do {
                            space()
                            val key = string()
                            if (values.containsKey(key)) fail("duplicate JSON object key")
                            space(); expect(':')
                            values[key] = value(depth + 1)
                            space()
                        } while (take(','))
                        expect('}')
                    }
                    JsonObject(values.toMap())
                }
                '-', in '0'..'9' -> {
                    val start = at
                    while (at < text.length && text[at] in "0123456789eE+.-") {
                        if (at - start >= budget.limits.maxNumberBytes) budget.resource("number byte limit exceeded")
                        at++
                    }
                    budget.spend(at - start)
                    JsonNumber.parse(text.substring(start, at))
                }
                else -> fail("invalid JSON value")
            }
        }
        private fun literal(expected: String) {
            budget.spend(expected.length)
            if (!text.startsWith(expected, at)) fail("invalid JSON literal")
            at += expected.length
        }
        private fun take(c: Char): Boolean {
            if (at < text.length && text[at] == c) { at++; budget.spend(1); return true }
            return false
        }
        private fun expect(c: Char) { if (!take(c)) fail("unexpected JSON delimiter") }
        private fun string(): String {
            expect('"')
            val result = StringBuilder()
            while (at < text.length) {
                budget.spend(1)
                val c = text[at++]
                when {
                    c == '"' -> return result.toString().also { utf8Size(it, budget.limits.maxBytes, budget.checkpoint) }
                    c < ' ' -> fail("control character in JSON string")
                    c == '\\' -> {
                        if (at == text.length) fail("unterminated escape")
                        budget.spend(1)
                        result.append(when (text[at++]) {
                            '"' -> '"'; '\\' -> '\\'; '/' -> '/'; 'b' -> '\b'; 'f' -> '\u000C'
                            'n' -> '\n'; 'r' -> '\r'; 't' -> '\t'
                            'u' -> {
                                if (text.length - at < 4) fail("short Unicode escape")
                                budget.spend(4)
                                var code = 0
                                repeat(4) {
                                    val digit = text[at++]
                                    if (digit !in "0123456789abcdefABCDEF") fail("invalid Unicode escape")
                                    code = code * 16 + digit.digitToInt(16)
                                }
                                code.toChar()
                            }
                            else -> fail("invalid JSON escape")
                        })
                    }
                    else -> result.append(c)
                }
            }
            fail("unterminated JSON string")
        }
    }

    private class Writer(private val budget: JsonBudget) {
        val out = StringBuilder()
        private var bytes = 0
        private fun append(text: String, size: Int = text.length) {
            if (size > budget.limits.maxBytes - bytes) budget.resource("JSON byte limit exceeded")
            bytes += size
            out.append(text)
        }
        fun value(value: JsonValue, depth: Int) {
            budget.value(depth)
            when (value) {
                JsonNull -> append("null")
                is JsonBoolean -> append(value.value.toString())
                is JsonNumber -> {
                    if (value.token.length > budget.limits.maxNumberBytes) budget.resource("number byte limit exceeded")
                    budget.spend(value.token.length); append(value.token)
                }
                is JsonString -> string(value.value)
                is JsonArray -> {
                    if (value.values.size > budget.limits.maxValues) budget.resource("JSON collection limit exceeded")
                    append("[")
                    value.values.forEachIndexed { index, item -> if (index > 0) append(","); value(item, depth + 1) }
                    append("]")
                }
                is JsonObject -> {
                    if (value.values.size > budget.limits.maxValues) budget.resource("JSON collection limit exceeded")
                    budget.spend(value.values.size)
                    append("{")
                    value.values.keys.sorted().forEachIndexed { index, key ->
                        if (index > 0) append(",")
                        string(key); append(":"); value(value.values.getValue(key), depth + 1)
                    }
                    append("}")
                }
            }
        }
        private fun string(value: String) {
            budget.spend(value.length)
            utf8Size(value, budget.limits.maxBytes, budget.checkpoint)
            append("\"")
            var index = 0
            while (index < value.length) {
                if (index % 1024 == 0) budget.checkpoint()
                val c = value[index++]
                when (c) {
                    '"' -> append("\\\""); '\\' -> append("\\\\")
                    '\n' -> append("\\n"); '\r' -> append("\\r"); '\t' -> append("\\t")
                    else -> when {
                        c < ' ' -> append("\\u" + c.code.toString(16).padStart(4, '0'))
                        c.isHighSurrogate() -> append("$c${value[index++]}", 4)
                        else -> append(c.toString(), if (c.code < 128) 1 else if (c.code < 2048) 2 else 3)
                    }
                }
            }
            append("\"")
        }
    }

    internal fun unicode(text: String) { utf8Size(text, Int.MAX_VALUE) {} }
    internal fun utf8Size(text: String, maximum: Int, checkpoint: () -> Unit): Int {
        if (text.length > maximum) throw JsonException("JSON byte limit exceeded", 0, JsonErrorKind.RESOURCE_LIMIT)
        var index = 0
        var bytes = 0
        while (index < text.length) {
            if (index % 1024 == 0) checkpoint()
            val c = text[index++]
            bytes += when {
                c.isHighSurrogate() -> {
                    if (index == text.length || !text[index++].isLowSurrogate()) throw JsonException("unpaired UTF-16 surrogate", index - 1)
                    4
                }
                c.isLowSurrogate() -> throw JsonException("unpaired UTF-16 surrogate", index - 1)
                c.code < 128 -> 1
                c.code < 2048 -> 2
                else -> 3
            }
            if (bytes > maximum) throw JsonException("JSON byte limit exceeded", 0, JsonErrorKind.RESOURCE_LIMIT)
        }
        checkpoint()
        return bytes
    }
}

// Normalization is proportional to token length, never to exponent magnitude.
internal data class ExactDecimal(val sign: Int, val digits: String, val exponent: BigInteger) : Comparable<ExactDecimal> {
    val integral: Boolean get() = sign == 0 || exponent.signum() >= 0
    override fun compareTo(other: ExactDecimal): Int {
        if (sign != other.sign) return sign.compareTo(other.sign)
        if (sign == 0) return 0
        val order = (exponent + BigInteger.valueOf(digits.length.toLong()))
            .compareTo(other.exponent + BigInteger.valueOf(other.digits.length.toLong()))
        if (order != 0) return sign * order
        for (i in 0 until maxOf(digits.length, other.digits.length)) {
            val difference = digits.getOrElse(i) { '0' }.compareTo(other.digits.getOrElse(i) { '0' })
            if (difference != 0) return sign * difference
        }
        return 0
    }
    fun multipleOf(other: ExactDecimal, spend: (Int) -> Unit): Boolean {
        check(other.sign > 0) { "compiled divisor must be positive" }
        if (sign == 0) return true
        val shift = exponent - other.exponent
        if (shift.signum() < 0) return false
        spend(1 + digits.length * other.digits.length / 64)
        val numerator = BigInteger(digits)
        val divisor = BigInteger(other.digits)
        var denominator = divisor / numerator.gcd(divisor)
        var twos = 0
        var fives = 0
        while (denominator.and(BigInteger.ONE) == BigInteger.ZERO) {
            spend(1 + denominator.bitLength() / 64)
            denominator = denominator.shiftRight(1); twos++
        }
        val five = BigInteger.valueOf(5)
        while (true) {
            spend(1 + denominator.bitLength() / 64)
            if (denominator.mod(five) != BigInteger.ZERO) break
            denominator /= five; fives++
        }
        return denominator == BigInteger.ONE && shift >= BigInteger.valueOf(maxOf(twos, fives).toLong())
    }
    companion object {
        fun parse(token: String): ExactDecimal {
            val negative = token.startsWith('-')
            val unsigned = if (negative) token.substring(1) else token
            val exponentAt = unsigned.indexOfAny(charArrayOf('e', 'E'))
            val mantissa = if (exponentAt < 0) unsigned else unsigned.substring(0, exponentAt)
            var exponent = if (exponentAt < 0) BigInteger.ZERO else BigInteger(unsigned.substring(exponentAt + 1))
            val dot = mantissa.indexOf('.')
            if (dot >= 0) exponent -= BigInteger.valueOf((mantissa.length - dot - 1).toLong())
            val digits = mantissa.replace(".", "").trimStart('0')
            if (digits.isEmpty()) return ExactDecimal(0, "0", BigInteger.ZERO)
            val trimmed = digits.trimEnd('0')
            exponent += BigInteger.valueOf((digits.length - trimmed.length).toLong())
            return ExactDecimal(if (negative) -1 else 1, trimmed, exponent)
        }
    }
}

File: crates/suspect-codegen/src/swift_sdk/json.swift
import Foundation

/// Finite exact-JSON limits. Each parse/write call owns its counters.
public struct JsonLimits: Sendable {
    public var maxBytes: Int
    public var maxDepth: Int
    public var maxNodes: Int
    public var maxNumberBytes: Int

    public init(maxBytes: Int = 8 * 1024 * 1024, maxDepth: Int = 128,
                maxNodes: Int = 100_000, maxNumberBytes: Int = 4096) {
        self.maxBytes = maxBytes
        self.maxDepth = maxDepth
        self.maxNodes = maxNodes
        self.maxNumberBytes = maxNumberBytes
    }

    func check() throws {
        guard (1...Int(Int32.max)).contains(maxBytes), (1...512).contains(maxDepth),
              (1...1_000_000).contains(maxNodes), (1...65_536).contains(maxNumberBytes) else {
            throw JsonError(.resourceLimit, "invalid JSON resource policy")
        }
    }
}

/// Syntax/representation failure, separate from a schema mismatch.
public struct JsonError: Error, Sendable, CustomStringConvertible {
    public enum Kind: String, Sendable { case syntax, resourceLimit, duplicateKey, invalidNumber, sourceBoundary, representation }
    public let kind: Kind
    public let message: String
    public let offset: Int?
    public let path: String
    public var description: String { "\(kind.rawValue): \(message) at \(path)\(offset.map { " byte \($0)" } ?? "")" }
    init(_ kind: Kind, _ message: String, offset: Int? = nil, path: String = "") {
        self.kind = kind; self.message = message; self.offset = offset; self.path = path
    }
}

/// An RFC 8259 number token. Construction validates grammar and never uses a
/// floating-point or Foundation NSNumber intermediate. Equality is token equality.
public struct JsonNumber: Sendable, Equatable, Hashable, ExpressibleByIntegerLiteral {
    public let raw: String
    public init(_ raw: String) throws {
        guard raw.utf8.count <= 65_536 else { throw JsonError(.resourceLimit, "number token exceeds 65536 bytes") }
        let bytes = Array(raw.utf8)
        var parser = JSONParser(bytes: bytes, limits: JsonLimits(maxNumberBytes: 65_536))
        _ = try parser.number()
        guard parser.at == bytes.count else { throw JsonError(.invalidNumber, "trailing bytes in number token", offset: parser.at) }
        self.raw = raw
    }
    public init(raw: String) throws { try self.init(raw) }
    public init(integerLiteral value: Int64) { self.raw = String(value) }
    public init(_ value: Int64) { self.raw = String(value) }
    // Only generated, compiler-checked literals and the strict parser use this.
    init(trusted raw: String) { self.raw = raw }
}

/// An exact mathematical integer, retaining spellings such as 1.00e+3.
public struct JsonInteger: Sendable, Equatable, Hashable, ExpressibleByIntegerLiteral {
    public let number: JsonNumber
    public var raw: String { number.raw }
    public init(_ raw: String) throws { try self.init(JsonNumber(raw)) }
    public init(_ number: JsonNumber) throws {
        guard ExactDecimal(number.raw).isIntegral else { throw JsonError(.invalidNumber, "integer value has a fractional component") }
        self.number = number
    }
    public init(integerLiteral value: Int64) { number = JsonNumber(value) }
    public init(_ value: Int64) { number = JsonNumber(value) }
    /// Checked narrowing. Out-of-range values throw; no zero substitution occurs.
    public func int64Value() throws -> Int64 {
        let exact = ExactDecimal(raw)
        if exact.sign == 0 { return 0 }
        guard exact.exponent.compare(BigSigned(19)) <= 0,
              let shift = exact.exponent.smallInt(), shift >= 0,
              exact.digits.count + shift <= 19 else { throw JsonError(.representation, "integer is outside Int64") }
        let text = (exact.sign < 0 ? "-" : "") + String(decoding: exact.digits.map { $0 + 48 }, as: UTF8.self) + String(repeating: "0", count: shift)
        guard let value = Int64(text) else { throw JsonError(.representation, "integer is outside Int64") }
        return value
    }
}

/// The only value of a source-declared null-only schema.
public struct JsonNull: Sendable, Equatable { public init() {} }

// Swift.String equality normalizes Unicode. JSON property identities must not:
// U+00E9 and U+0065 U+0301 are different names under JSON Schema.
struct JsonKey: Sendable, Hashable {
    let text: String
    init(_ text: String) { self.text = text }
    static func == (a: Self, b: Self) -> Bool { a.text.utf8.elementsEqual(b.text.utf8) }
    func hash(into hasher: inout Hasher) {
        hasher.combine(text.utf8.count)
        for byte in text.utf8 { hasher.combine(byte) }
    }
}

/// A value-semantic JSON object with exact Unicode key identity. Keys are sorted
/// by UTF-8 for deterministic encoding; mutation preserves every distinct name.
public struct JsonObject<Value: Sendable>: Sendable {
    private var storage: [JsonKey: Value] = [:]
    public init() {}
    public init(_ members: [(String, Value)]) throws {
        for (key, value) in members {
            guard storage.updateValue(value, forKey: JsonKey(key)) == nil else {
                throw JsonError(.duplicateKey, "duplicate decoded object key")
            }
        }
    }
    init(trusted members: [(String, Value)]) {
        for (key, value) in members { storage[JsonKey(key)] = value }
    }
    public subscript(key: String) -> Value? {
        get { storage[JsonKey(key)] }
        set { storage[JsonKey(key)] = newValue }
    }
    public var count: Int { storage.count }
    public var isEmpty: Bool { storage.isEmpty }
    public var members: [(key: String, value: Value)] {
        storage.map { (key: $0.key.text, value: $0.value) }.sorted { $0.key.utf8.lexicographicallyPrecedes($1.key.utf8) }
    }
    public var keys: [String] { members.map(\.key) }
}
extension JsonObject: Equatable where Value: Equatable {
    public static func == (a: Self, b: Self) -> Bool { a.storage == b.storage }
}

/// Source-preserving JSON values. Numeric tokens and Unicode scalar sequences
/// survive round trips; whitespace, key order and string escape spellings may change.
public indirect enum JsonValue: Sendable, Equatable {
    case null
    case bool(Bool)
    case number(JsonNumber)
    case string(String)
    case array([JsonValue])
    case object(JsonObject<JsonValue>)

    public static func == (a: Self, b: Self) -> Bool {
        switch (a, b) {
        case (.null, .null): return true
        case let (.bool(a), .bool(b)): return a == b
        case let (.number(a), .number(b)): return a == b
        case let (.string(a), .string(b)): return a.utf8.elementsEqual(b.utf8)
        case let (.array(a), .array(b)): return a == b
        case let (.object(a), .object(b)): return a == b
        default: return false
        }
    }
    public static func parse(_ data: Data, limits: JsonLimits = .init()) throws -> Self {
        try limits.check()
        guard data.count <= limits.maxBytes else { throw JsonError(.resourceLimit, "JSON input byte limit exceeded") }
        var parser = JSONParser(bytes: Array(data), limits: limits)
        let result = try parser.value(depth: 0)
        parser.space()
        guard parser.at == data.count else { throw JsonError(.syntax, "trailing bytes after JSON value", offset: parser.at) }
        return result
    }
    public static func parse(_ text: String, limits: JsonLimits = .init()) throws -> Self {
        try limits.check()
        guard text.utf8.count <= limits.maxBytes else { throw JsonError(.resourceLimit, "JSON input byte limit exceeded") }
        return try parse(Data(text.utf8), limits: limits)
    }
    public func encoded(limits: JsonLimits = .init()) throws -> Data {
        try limits.check()
        var writer = JSONWriter(limits: limits)
        try writer.value(self, depth: 0)
        return Data(writer.bytes)
    }
    public func encodedString(limits: JsonLimits = .init()) throws -> String {
        String(decoding: try encoded(limits: limits), as: UTF8.self)
    }
    var isNull: Bool { if case .null = self { return true }; return false }
    var kind: String {
        switch self {
        case .null: return "null"
        case .bool: return "boolean"
        case .number: return "number"
        case .string: return "string"
        case .array: return "array"
        case .object: return "object"
        }
    }
}

/// Compatibility spelling for the exact JSON value domain.
public typealias ExactValue = JsonValue

struct JSONParser {
    let bytes: [UInt8]
    let limits: JsonLimits
    var at = 0
    var nodes = 0

    mutating func space() {
        while at < bytes.count && [UInt8(32), 9, 10, 13].contains(bytes[at]) { at += 1 }
    }
    func error(_ message: String) -> JsonError { JsonError(.syntax, message, offset: at) }
    mutating func take(_ byte: UInt8) -> Bool {
        guard at < bytes.count && bytes[at] == byte else { return false }
        at += 1; return true
    }
    mutating func value(depth: Int) throws -> JsonValue {
        guard depth < limits.maxDepth && nodes < limits.maxNodes else { throw JsonError(.resourceLimit, "JSON depth or node limit exceeded", offset: at) }
        nodes += 1; space()
        guard at < bytes.count else { throw error("expected JSON value") }
        switch bytes[at] {
        case 110: try literal("null"); return .null
        case 116: try literal("true"); return .bool(true)
        case 102: try literal("false"); return .bool(false)
        case 34: return .string(try string())
        case 91:
            at += 1; space()
            var items: [JsonValue] = []
            if take(93) { return .array(items) }
            while true {
                items.append(try value(depth: depth + 1)); space()
                if take(93) { return .array(items) }
                guard take(44) else { throw error("expected ',' or ']'") }
            }
        case 123:
            at += 1; space()
            var object = JsonObject<JsonValue>()
            if take(125) { return .object(object) }
            while true {
                space()
                guard at < bytes.count && bytes[at] == 34 else { throw error("expected object key") }
                let key = try string(); space()
                guard take(58) else { throw error("expected ':'") }
                guard object[key] == nil else { throw JsonError(.duplicateKey, "duplicate decoded object key", offset: at) }
                object[key] = try value(depth: depth + 1); space()
                if take(125) { return .object(object) }
                guard take(44) else { throw error("expected ',' or '}'") }
            }
        default: return .number(try number())
        }
    }
    mutating func literal(_ text: String) throws {
        for byte in text.utf8 { guard take(byte) else { throw error("invalid literal") } }
    }
    mutating func number() throws -> JsonNumber {
        let start = at
        _ = take(45)
        guard at < bytes.count else { throw error("incomplete number") }
        if take(48) {
            if at < bytes.count && (48...57).contains(bytes[at]) { throw error("leading zero in number") }
        } else {
            guard (49...57).contains(bytes[at]) else { throw error("invalid number") }
            while at < bytes.count && (48...57).contains(bytes[at]) { at += 1 }
        }
        if take(46) {
            let digits = at
            while at < bytes.count && (48...57).contains(bytes[at]) { at += 1 }
            guard at > digits else { throw error("fraction requires digits") }
        }
        if take(101) || take(69) {
            if !take(43) { _ = take(45) }
            let digits = at
            while at < bytes.count && (48...57).contains(bytes[at]) { at += 1 }
            guard at > digits else { throw error("exponent requires digits") }
        }
        guard at - start <= limits.maxNumberBytes else { throw JsonError(.resourceLimit, "number byte limit exceeded", offset: start) }
        return JsonNumber(trusted: String(decoding: bytes[start..<at], as: UTF8.self))
    }
    mutating func hex() throws -> UInt32 {
        var value: UInt32 = 0
        for _ in 0..<4 {
            guard at < bytes.count else { throw error("incomplete Unicode escape") }
            let byte = bytes[at]; at += 1
            let digit: UInt32
            switch byte {
            case 48...57: digit = UInt32(byte - 48)
            case 65...70: digit = UInt32(byte - 55)
            case 97...102: digit = UInt32(byte - 87)
            default: throw error("invalid Unicode escape")
            }
            value = value * 16 + digit
        }
        return value
    }
    mutating func string() throws -> String {
        guard take(34) else { throw error("expected string") }
        var out: [UInt8] = []
        while at < bytes.count {
            let byte = bytes[at]; at += 1
            if byte == 34 {
                guard let text = String(bytes: out, encoding: .utf8) else { throw error("invalid UTF-8") }
                return text
            }
            if byte == 92 {
                guard at < bytes.count else { throw error("incomplete escape") }
                let escaped = bytes[at]; at += 1
                switch escaped {
                case 34, 47, 92: out.append(escaped)
                case 98: out.append(8)
                case 102: out.append(12)
                case 110: out.append(10)
                case 114: out.append(13)
                case 116: out.append(9)
                case 117:
                    var scalar = try hex()
                    if (0xd800...0xdbff).contains(scalar) {
                        guard take(92), take(117) else { throw error("high surrogate requires a low surrogate") }
                        let low = try hex()
                        guard (0xdc00...0xdfff).contains(low) else { throw error("invalid low surrogate") }
                        scalar = 0x10000 + (scalar - 0xd800) * 0x400 + low - 0xdc00
                    }
                    guard let unicode = Unicode.Scalar(scalar) else { throw error("unpaired surrogate") }
                    out.append(contentsOf: String(unicode).utf8)
                default: throw error("invalid escape")
                }
            } else {
                guard byte >= 32 else { throw error("unescaped control character") }
                out.append(byte)
            }
        }
        throw error("unterminated string")
    }
}

struct JSONWriter {
    let limits: JsonLimits
    var bytes: [UInt8] = []
    var nodes = 0
    mutating func append<S: Sequence>(_ values: S) throws where S.Element == UInt8 {
        for byte in values {
            guard bytes.count < limits.maxBytes else { throw JsonError(.resourceLimit, "JSON output byte limit exceeded") }
            bytes.append(byte)
        }
    }
    mutating func byte(_ value: UInt8) throws { try append(CollectionOfOne(value)) }
    mutating func quoted(_ text: String) throws {
        try byte(34)
        for scalar in text.unicodeScalars {
            switch scalar.value {
            case 34: try append("\\\"".utf8)
            case 92: try append("\\\\".utf8)
            case 8: try append("\\b".utf8)
            case 9: try append("\\t".utf8)
            case 10: try append("\\n".utf8)
            case 12: try append("\\f".utf8)
            case 13: try append("\\r".utf8)
            case 0..<32:
                let hex = String(scalar.value, radix: 16)
                try append(("\\u" + String(repeating: "0", count: 4 - hex.count) + hex).utf8)
            default: try append(String(scalar).utf8)
            }
        }
        try byte(34)
    }
    mutating func value(_ value: JsonValue, depth: Int) throws {
        guard depth < limits.maxDepth && nodes < limits.maxNodes else { throw JsonError(.resourceLimit, "JSON depth or node limit exceeded") }
        nodes += 1
        switch value {
        case .null: try append("null".utf8)
        case .bool(let b): try append((b ? "true" : "false").utf8)
        case .number(let n):
            guard n.raw.utf8.count <= limits.maxNumberBytes else { throw JsonError(.resourceLimit, "number byte limit exceeded") }
            try append(n.raw.utf8)
        case .string(let s): try quoted(s)
        case .array(let items):
            try byte(91)
            for (i, item) in items.enumerated() {
                if i > 0 { try byte(44) }
                try self.value(item, depth: depth + 1)
            }
            try byte(93)
        case .object(let object):
            guard object.count <= limits.maxNodes - nodes else { throw JsonError(.resourceLimit, "JSON object exceeds remaining node limit") }
            try byte(123)
            for (i, member) in object.members.enumerated() {
                if i > 0 { try byte(44) }
                try quoted(member.key); try byte(58)
                try self.value(member.value, depth: depth + 1)
            }
            try byte(125)
        }
    }
}
